### PR TITLE
cleanup(framework): return earlier if all Score plugins are skipped

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

do an earlier return when all Score plugins returned Skip in PreScore.

<details><summary>This PR originally had the fix for panic bug</summary><div>

The scheduler panics when all these conditions are satisfied:
- some PreScore plugins which returned Skip, but are disabled in the Score extension point
- "the number of PreScore plugins that returned Skip and disabled in the Score" is more than the number of all Score plugins enabled in the scheduler

You can reproduce the panic with newly added UT like the following.

```
Running tool: /home/gitpod/go/bin/go test -timeout 30s -run ^TestRunScorePlugins$ k8s.io/kubernetes/pkg/scheduler/framework/runtime

I0527 12:20:50.858222  101963 framework.go:509] "MultiPoint plugin is explicitly re-configured; overriding" plugin="score-plugin-1"
--- FAIL: TestRunScorePlugins (0.00s)
    --- FAIL: TestRunScorePlugins/PreScore_returns_Skip_but_isn't_registered_in_Score (0.00s)
panic: runtime error: makeslice: cap out of range [recovered]
	panic: runtime error: makeslice: cap out of range
```

</div></details>

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

<details><summary>This PR originally had the fix for panic bug</summary><div>=

@kubernetes/sig-scheduling-leads 
In my opinion, we don't need to cherry-pick this into v1.27, but please anyone double-check.

This shouldn't break the default Kubernetes scheduler because any in-tree Score plugins don't return Skip in v1.27 in the first place. ([ref](https://github.com/kubernetes/kubernetes/issues/115745))
It might break the custom scheduler, for example, the scheduler with some custom PreScore plugins which may return Skip, but are disabled in the Score extension point. Also, to see this bug, "the number of PreScore plugins that returned Skip and disabled in the Score" should be more than the number of all Score plugins enabled in the scheduler, which is an extremely minor case.

</div></details>

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
/priority important-soon
/triage accepted

As described, it's not critical. But, it's a blocker to implement Skip in Score plugins because it's breaking the e2e test in the PR of Skip implementation (e.g., https://github.com/kubernetes/kubernetes/pull/117024).

File: pkg/scheduler/framework/runtime/framework.go [hunk 1]
```diff
@@ -1014,49 +1014,57 @@ func (f *frameworkImpl) RunScorePlugins(ctx context.Context, state *framework.Cy
 		metrics.FrameworkExtensionPointDuration.WithLabelValues(metrics.Score, status.Code().String(), f.profileName).Observe(metrics.SinceInSeconds(startTime))
 	}()
 	allNodePluginScores := make([]framework.NodePluginScores, len(nodes))
-	numPlugins := len(f.scorePlugins) - state.SkipScorePlugins.Len()
```
Comment (sanposhiho) [inline pkg/scheduler/framework/runtime/framework.go:1017]: The trigger of this bug is that `numPlugins` becomes less than 0.

**Comment by AxeZhan** [comment]
I ran into this before in some tests, but is there really a scenario that a plugin is enabled for prescore and disabled for score in reality?

File: pkg/scheduler/framework/runtime/framework.go [hunk 2]
```diff
@@ -1014,49 +1014,57 @@ func (f *frameworkImpl) RunScorePlugins(ctx context.Context, state *framework.Cy
...
+	pluginToNodeScores := make(map[string]framework.NodeScoreList)
 	for _, pl := range f.scorePlugins {
 		if state.SkipScorePlugins.Has(pl.Name()) {
 			continue
 		}
 		plugins = append(plugins, pl)
 		pluginToNodeScores[pl.Name()] = make(framework.NodeScoreList, len(nodes))
 	}
+
+	if len(plugins) < 0 {
```
Comment (AxeZhan) [inline pkg/scheduler/framework/runtime/framework.go:1027]: <= ?

**Comment by sanposhiho** [comment]
> I ran into this before in some tests, but is there really a scenario that a plugin is enabled for prescore and disabled for score in reality?

I'd say 100% No if no one made mistakes in the scheduler component config. This fix is for mistakers.

/remove-priority important-soon

I'll fix the e2e test side to unblock the Skip PR.

**Comment by sanposhiho** [comment]
But, on second thought, we may just want to accept this bug because this PR will remove pre-allocation for `pluginToNodeScores` and `plugins`. 🤔

**Comment by sanposhiho** [comment]
> But, on second thought, we may just want to accept this bug because this PR will remove pre-allocation for pluginToNodeScores and plugins. 🤔

Let's go with this thought unless other people want this bug fix.
I'll convert this PR to the cleanup one. Feel free to give your thought if you think that's worthwhile to have.

/remove-kind bug
/retitle cleanup(framework): return earlier if all Score plugins are skipped

**Review by Tusenka — APPROVED** [review]

**Comment by kerthcet** [comment]
Can you fix the test?

**Comment by sanposhiho** [comment]
It's in icebox state in my todo list. I don't think I can take time to take care of it for now. Anyone feel free to raise a similar PR. (or maybe future me will)

/close

**Pull request closed without merging**